GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GitFilesTests.cs

```diff
@@ -353,44 +353,97 @@ public void SupersededFileAddedToModifiedPathsAndSkipWorktreeBitCleared()
-            fileOutsideRepo.ShouldNotExistOnDisk(this.fileSystem);
+            fileOutsideRepo.ShouldNotExistOnDisk(this.fileSystem);
             this.Enlistment.WaitForBackgroundOperations();
             GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
-        }
-
+        }
+
         [TestCase, Order(17)]
         public void FileMovedFromInsideRepoToOutside()
         {
             string fileName = "InsideRepoToOutside.txt";
-            string fileInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
-            this.fileSystem.WriteAllText(fileInsideRepo, "Contents for the new file");
-            fileInsideRepo.ShouldBeAFile(this.fileSystem);
+            string fileInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
+            this.fileSystem.WriteAllText(fileInsideRepo, "Contents for the new file");
+            fileInsideRepo.ShouldBeAFile(this.fileSystem);
             this.Enlistment.WaitForBackgroundOperations();
             GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
-
-            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
-            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
+
+            string fileMovedOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.MoveFile(fileInsideRepo, fileMovedOutsideRepo);
             fileInsideRepo.ShouldNotExistOnDisk(this.fileSystem);
-            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
-            this.Enlistment.WaitForBackgroundOperations();
+            fileMovedOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
             GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
         }
 
+        [TestCase, Order(18)]
+        public void HardlinkFromOutsideRepoToInside()
+        {
+            string fileName = "OutsideRepoFileForHardlink.txt";
+            string fileOutsideRepo = Path.Combine(this.Enlistment.EnlistmentRoot, fileName);
+            this.fileSystem.WriteAllText(fileOutsideRepo, "Contents for the new file");
+            fileOutsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldNotContain(this.Enlistment, this.fileSystem, fileName);
+
+            string fileHardlinkedInsideRepo = this.Enlistment.GetVirtualPathTo(fileName);
+            this.fileSystem.CreateHardLink(fileHardlinkedInsideRepo, fileOutsideRepo);
+            fileHardlinkedInsideRepo.ShouldBeAFile(this.fileSystem);
+            this.Enlistment.WaitForBackgroundOperations();
+            GVFSHelpers.ModifiedPathsShouldContain(this.Enlistment, this.fileSystem, fileName);
+        }
+
+        [TestCase, Order(19)]
+        public void HardlinkFromInsideRepoToInside()

```

Comment:
This test name says `FromInsideRepoToInside` but it looks like one of the files being linked is outside of the repo.